### PR TITLE
Add mozilla.org - from code to production post

### PR DESCRIPTION
**Pull request by alexgibson**
@pmac - I made a first pass at laying out the post structure. I'm on PTO next week, but feel free to add / change anything if you have any time. Just ideas currently, nothing concrete.

Note: this is in `_drafts` so you'll need to start the server with the `--drafts` flag to see it.

### CODE REVIEW

**Comment by jpetto** [comment]
Should this be a page instead of a post? I feel like this should be a living document directly linked off the home page. 😁

**Comment by alexgibson** [comment]
> Should this be a page instead of a post? I feel like this should be a living document directly linked off the home page. 😁

Interesting idea 👍

File: _drafts/bedrock-code-to-production.md [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+### Special considerations for download pages
+
+For critical download pages such as our example page, we still need to ensure that users can successfully download Firefox, even in older browsers. When the user clicks the download button, they still need to receive the correct file for their operating system and language.
+
+Because of this added support requirement, download pages may require additional testing and QA. For really high traffic pages such as `/firefox/new/`, we may still make extra effort to provide a higher level of CSS support to older browsers. The pages won't look look exactly the same, but they should still degrade gracefully.
+
+### Localization (l10n)
+
+Many mozilla.org pages are translated into multiple languages. While much of our marketing focus is primarily for English language locales first and foremost, non-English traffic actually makes up around 60% (TODO: is this accurate?) of our overall site traffic. Unless there is a specific request for a page to be in English only, we create all pages assuming that they will be translated should our volunteer community wish to do the work.
```
Comment (pmac) [inline _drafts/bedrock-code-to-production.md]: I believe the 60% number is close enough to say "around ...". Maybe check with `pg` before publishing?

File: _drafts/bedrock-code-to-production.md [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+layout: post
+title: Bedrock - from code to production
+author: Alex Gibson &amp; Paul McLanahan
+---
+
+## Introduction
+
+Bedrock is the code name of [www.mozilla.org](https://www.mozilla.org/). It is shiny, awesome, and [open source](https://github.com/mozilla/bedrock/). Mozilla.org is the site where everyone comes to download Firefox, and is also the main public facing web property for Mozilla as an organization. It represents [who we are](https://www.mozilla.org/en-US/about/) and [what we do](https://www.mozilla.org/en-US/mission/).
+
+Bedrock is a large [Django](https://www.djangoproject.com/) app that has many moving parts, with web pages translated into [many](https://www.mozilla.org/de/) [different](https://www.mozilla.org/fr/) [languages](https://www.mozilla.org/es-ES/) (over X locales!). This post aims to follow a piece of code through the development lifecycle; from bug, to pull request, to code review, localization, and finally pushed live to production. Hopefully it may prove insightful and give some things to think about when requesting changes or creating new content.
```
Comment (pmac) [inline _drafts/bedrock-code-to-production.md]: maybe say "(99 at time of writing!)

**Comment by pmac** [comment]
> Should this be a page instead of a post? I feel like this should be a living document directly linked off the home page. 😁

I dunno... I like having this be a description of the process as at this moment in time. If we want a living version we should probably adapt it for the bedrock docs. I could be convinced either way though I think.

**Comment by alexgibson** [comment]
> I dunno... I like having this be a description of the process as at this moment in time. If we want a living version we should probably adapt it for the bedrock docs. I could be convinced either way though I think.

Yeah, I've been giving this some thought as I write things. One thing I'm hoping for this to do is to appeal to a wider circle of people (e.g. project managers) than say, a piece of developer documentation would do. I'm trying to write it more as a high-level tutorial for our processes, rather than something I think should live in the bedrock docs. Personally, I'd be OK with this going out of date (seeing the time of publishing is included in each post), but if we wanted to adapt a separate version for the bedrock docs we could certainly do that too.

**Comment by alexgibson** [comment]
@pmac - OK, I think I'm pretty much done with the bits except where I think you're knowledge would be appreciated. Feel free to change anything though!

**Comment by jpetto** [comment]
> I dunno... I like having this be a description of the process as at this moment in time. If we want a living version we should probably adapt it for the bedrock docs. I could be convinced either way though I think.

I'm on-board with having this version be for a moment in time, but keeping an up-to-date version in the docs. 👍

**Comment by alexgibson** [comment]
@ejregithub - if you have some time I'd very much be interested in your feedback on what I have so far here (this blog post will eventually form the basis for the open session we're planning on doing). I was wondering if it might be worth including a part near the start about the process for requesting new content (via your form or similar).

**Comment by ejregithub** [comment]
I think that work request intake form would unnecessarily complicate the blog. plus, that's intended  for more complex projects and/or a means to weed out poorly thought-out requests for the same. Going straight to filing a bug is appropriate to the intent of this post. In bug triage we can determine if sending the person(s) who submitted it if they should be guided to the work request intake form to better understand their asks.

+1 to this launching as a blog post
+1 to the contents of the post becoming a living doc on mozilla.org and being formatted as a 'best practices' document (we can then add a section about the work request intake form).

**Comment by pmac** [comment]
I've added some updates to the branch. One is just a script that will run the site in dev-mode using docker. So if you have docker installed it should Just Work™️. The other is changes to the post.

**Note:** I had a bit of a hard time with the merge because the original commit I based my work on was amended and force-pushed. Please in future simply push new commits if the branch is meant for collaboration. It makes contribution from multiple people much easier. Also make sure I didn't hose any of your recent changes in said merge resolution.

**Comment by alexgibson** [comment]
> Note: I had a bit of a hard time with the merge because the original commit I based my work on was amended and force-pushed. Please in future simply push new commits if the branch is meant for collaboration. It makes contribution from multiple people much easier. Also make sure I didn't hose any of your recent changes in said merge resolution.

Ah sorry about that. I thought I was safe when I last updated my commit, but I didn't realize you had started on this at the time. Will double check in future.

**Comment by alexgibson** [comment]
@pmac - this all looks great 👍

I added a short conclusion paragraph at the end, and fixed a couple of typos I spotted. I think we can probably call this complete? 🎉 

Before we merge, it would be good for either @jpetto or @craigcook to give this a read through?

**Comment by pmac** [comment]
Love it! Thanks Alex!

**Comment by ejregithub** [comment]
This is good work.  Very approachable, imparting enough information to convey what 'a thing' does without getting overly technical (as was the intent here, I believe!).

File: _drafts/bedrock-code-to-production.md [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+- Pre-built for [localization](http://bedrock.readthedocs.io/en/latest/l10n.html) (l10n). All pages are coded with strings wrapped ready for l10n. This means that copy and visuals should be suitable for different languages and designed to work well with variable length strings.
+- Built-in platform detection for Firefox downloads. We go to great lengths to try and make sure users get the correct binary file for their operating system and language when clicking on a download button. All pages get this logic for free should they need it.
+- Privacy friendly. We work hard to make sure all our pages respect user privacy as much as possible. We support [Do Not Track](https://en.wikipedia.org/wiki/Do_Not_Track) (DNT), so users with this enabled are not tracked in Google Analytics or entered into [Firefox Desktop Attribution](https://support.mozilla.org/t5/Protect-your-privacy/About-Firefox-Desktop-Attribution/ta-p/1361183). They will also not be entered into A/B experiments using libraries such as [Traffic Cop](http://bedrock.readthedocs.io/en/latest/mozilla-traffic-cop.html).
```
Comment (jpetto) [inline _drafts/bedrock-code-to-production.md]: Update Traffic Cop link to new repo?

File: _drafts/bedrock-code-to-production.md [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+  - Branch name must start with `demo__` and then be any letters, numbers, and dashes (e.g. demo__the-dude)
+2. Profit
+
+That's really it. Jenkins will build the demo and if it is successful a notification in our IRC channel will inform
+the developer of their new demo's URL. Feel free to check out
+[how that works](https://github.com/mozilla/bedrock/blob/139e8e589b2ebe391cb159068152408e9f090687/Jenkinsfile#L163-L203) if you're curious.
+
+### A/B testing
+
+For A/B testing bedrock has two options available, [Traffic Cop](http://bedrock.readthedocs.io/en/latest/mozilla-traffic-cop.html) and [Optimizely](https://www.optimizely.com/). [Jon](https://github.com/jpetto/) wrote a great blog post on the pros and cons of using each solution so we won't repeat them here. You should [go read it]({{ site.baseurl }}/2017/01/16/traffic-cop/).
```
Comment (jpetto) [inline _drafts/bedrock-code-to-production.md]: Link to new Traffic Cop repo?

File: _drafts/bedrock-code-to-production.md [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+
+### Analytics
+
+Analytics on bedrock are provided using [Google Tag Manager](https://www.google.com/analytics/tag-manager/) (GTM). Most of bedrocks shared components and pages are pre-built for analytics using common [event handlers and data attributes](http://bedrock.readthedocs.io/en/latest/analytics.html#gtm-listeners-data-attributes). We can also create and fire [custom events](http://bedrock.readthedocs.io/en/latest/analytics.html#datalayer-push-example) as required.
```
Comment (jpetto) [inline _drafts/bedrock-code-to-production.md]: Small nit - missing apostrophe. Should be `bedrock's shared components` (not `bedrocks`).

**Comment by jpetto** [comment]
Found one missing apostrophe, and suggested updating Traffic Cop links, but otherwise looks great. Amazing work @alexgibson & @pmac!

**Comment by alexgibson** [comment]
Thanks @jpetto 👍 

Updated with suggested fixes, also moved file out of `_drafts` and into `_posts`. I think we're good to publish this one? 😄

**Comment by jpetto** [comment]
:shipit:

**Comment by metadave** [comment]
very nice!

**Pull request merged**